File: code/R/H2O-AutoEncoder-Ecg.R
###: Importing H2O and other required libraries
library(h2o)


###: Initializing H2O cluster
h2o.init()


###: Importing dataset
train_ecg = h2o.importFile("https://raw.githubusercontent.com/Avkash/mldl/master/data/ecg_discord_train.csv")
test_ecg = h2o.importFile("https://raw.githubusercontent.com/Avkash/mldl/master/data/ecg_discord_test.csv")

###: Understand training dataset
h2o.describe(train_ecg)

###: transpose the frame to have the time series as a single colum to plot
train_ecg_transpose = t(as.data.frame(train_ecg))
plot(train_ecg_transpose,  x = 1:length(train_ecg_transpose),  col="blue")

###: Using Configured Deep Learning to train neural network
auto_encoder_model = h2o.deeplearning( 
  activation="Tanh", 
  hidden=c(50,50), 
  autoencoder = TRUE,
  l1=1e-5,
  score_interval=0,
  epochs=100,
  x=names(train_ecg), training_frame=train_ecg
)

###: Getting model details:
auto_encoder_model

###: Computing reconstruction error with the Anomaly detection function. 
###: This is the Mean Square Error between output and input layers.

reconstruction_error = h2o.anomaly(auto_encoder_model, test_ecg)

###: Now the question is: Which of the test_ecg time series are most likely an anomaly? 
###: Lets select the top N that have high error rate
###:Pulling reconstruction error data into R and plotting to find outliers (last 3 heartbeats)
reconstruction_error <- as.data.frame(reconstruction_error)
reconstruction_error
plot.ts(reconstruction_error)

###: Listing anomalies based on reconstucted MSE values:
anomalies = reconstruction_error[reconstruction_error['Reconstruction.MSE'] > 1.0]
anomalies
###: Plotting test data without anomalies

# Note: Testing = Reconstructing the test dataset
test_recon =  h2o.predict(auto_encoder_model, test_ecg) 
head(test_recon)

File: code/scala/H2O-AutoEncoder-ECG.scala
///### Importing key libraries:
import org.apache.spark.h2o._
import org.apache.spark.SparkFiles
import org.apache.spark.examples.h2o._
import org.apache.spark.sql.{DataFrame, SQLContext}
import water.Key
import water.support.SparkContextSupport.addFiles
import water.support.H2OFrameSupport._
import java.io.File

//### Creating H2O Context:
val h2oContext = H2OContext.getOrCreate(sc)
import h2oContext._
import h2oContext.implicits._



val ecgTrainData = new H2OFrame(new File("/Users/avkashchauhan/learn/customers/ecg_discord_train.csv"))
val ecgTestData = new H2OFrame(new File("/Users/avkashchauhan/learn/customers/ecg_discord_test.csv"))

import _root_.hex.deeplearning
import _root_.hex.deeplearning.DeepLearning
import _root_.hex.deeplearning.DeepLearningModel
import _root_.hex.deeplearning.DeepLearningModelInfo
import _root_.hex.deeplearning.DeepLearningModel.DeepLearningParameters
import _root_.hex.deeplearning.DeepLearningModel.DeepLearningParameters.Activation


val dlParams = new DeepLearningParameters()

dlParams._train = ecgTrainData
dlParams._hidden = Array[Int](100, 100)
dlParams._epochs = 5
dlParams._activation = Activation.RectifierWithDropout
dlParams._autoencoder = true
dlParams._destination_key = Key.make("dlEncoderModel.hex").asInstanceOf[water.Key[Frame]]


// Create a job
val dl = new DeepLearning(dlParams)
val dlEncoderModel = dl.trainModel.get

// Reconstruction error per feature using test Dataset:
val reconstruction_error =  dlEncoderModel.scoreAutoEncoder(ecgTestData, Key.make(), true)

// List of names into reconstruction_error
reconstruction_error.names

// Reconstruction error per feature using test Dataset:
val reconstruction_error =  dlEncoderModel.scoreAutoEncoder(ecgTestData, Key.make(), false)
val reconstruction_error_vec = reconstruction_error.anyVec()

// Calculate mean
val error_mean = reconstruction_error_vec.mean

// Checking MSE from the model
val model_mse = dlEncoderModel.mse()


